zerver/views/message_fetch.py

```diff
@@ -106,19 +106,40 @@ def get_messages_backend(
     *,
     anchor_val: Annotated[str | None, ApiParamConfig("anchor")] = None,
     include_anchor: Json[bool] = True,
-    num_before: Json[NonNegativeInt],
-    num_after: Json[NonNegativeInt],
+    num_before: Json[NonNegativeInt] = 0,
+    num_after: Json[NonNegativeInt] = 0,
     narrow: Json[list[NarrowParameter] | None] = None,
     use_first_unread_anchor_val: Annotated[
         Json[bool], ApiParamConfig("use_first_unread_anchor")
     ] = False,
     client_gravatar: Json[bool] = True,
     apply_markdown: Json[bool] = True,
+    message_ids_to_fetch: Annotated[
+        Json[list[NonNegativeInt] | None], ApiParamConfig("message_ids")
+    ] = None,
 ) -> HttpResponse:
+    if message_ids_to_fetch is None:
+        message_ids_to_fetch = []
+    # User requested a list of message ids to fetch.
+    is_message_ids_pre_specified = len(message_ids_to_fetch) > 0
+    # User has to either provide message_ids or both num_before and num_after.
+    if (
+        num_before or num_after or anchor_val is not None or use_first_unread_anchor_val
+    ) and is_message_ids_pre_specified:
+        raise JsonableError(
+            _("`message_ids` are mutually exclusive with `num_before`, `num_after` and `anchor`")
```

Comment:
You can use `IncompatibleParametersError` to avoid extra overly specific translated strings.

Reply:
Done